Selective meta programming preparation for prim ops

iseeyuan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43540 Selective meta programming preparation for prim ops**

Differential Revision: [D22618309](https://our.internmc.facebook.com/intern/diff/D22618309/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22618309/)!

iseeyuan committed:
Selective meta programming preparation for prim ops



Differential Revision: [D22618309](https://our.internmc.facebook.com/intern/diff/D22618309/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22618309/)!

[ghstack-poisoned]

iseeyuan committed:
Update on "Selective meta programming preparation for prim ops"


Differential Revision: [D22618309](https://our.internmc.facebook.com/intern/diff/D22618309/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22618309/)!

[ghstack-poisoned]

iseeyuan committed:
Update on "Selective meta programming preparation for prim ops"


Differential Revision: [D22618309](https://our.internmc.facebook.com/intern/diff/D22618309/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22618309/)!

[ghstack-poisoned]

ezyang commented:
Just one quick note: if we are going to be OK with generating a header file that says what is enabled or not, we can probably simplify the template metaprogramming quite a bit; e.g., instead of having to shove everything in a comma-separated string, we can just directly encode it as a constexpr array.

iseeyuan commented:
> Just one quick note: if we are going to be OK with generating a header file that says what is enabled or not, we can probably simplify the template metaprogramming quite a bit; e.g., instead of having to shove everything in a comma-separated string, we can just directly encode it as a constexpr array.

Yes. Currently the header is only included when needed (internally). The logic of #ifdef is kept to check if we need to do selective. Probably later we could always include the header and don't do selective if it's empty?

iseeyuan committed:
Update on "Selective meta programming preparation for prim ops"


Differential Revision: [D22618309](https://our.internmc.facebook.com/intern/diff/D22618309/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22618309/)!

[ghstack-poisoned]

iseeyuan committed:
Update on "Selective meta programming preparation for prim ops"


Differential Revision: [D22618309](https://our.internmc.facebook.com/intern/diff/D22618309/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22618309/)!

[ghstack-poisoned]
